File: corehydror/tests/testthat/test-gof.R
test_that("goodness_of_fit returns every metric, and subsets agree with the whole", {
  obs <- c(2, 4, 6, 8, 10)
  mod <- c(2.2, 3.9, 6.4, 7.5, 10.1)
  all <- goodness_of_fit(obs, mod)
  expect_length(all, 17L)
  expect_true(all(c("nse", "kge", "rsr", "ve") %in% names(all)))
  expect_equal(goodness_of_fit(obs, mod, metrics = "nse")[["nse"]], all[["nse"]])
})

test_that("a subset that excludes mape succeeds on a zero-containing series", {
  obs <- c(0, 2, 4)
  mod <- c(0.1, 2.1, 3.9)
  expect_error(goodness_of_fit(obs, mod), "zero")
  result <- goodness_of_fit(obs, mod, metrics = "mse")
  expect_true(is.finite(result[["mse"]]))
})

test_that("a subset returns the same values as the full call, in the requested order", {
  obs <- c(2, 4, 6, 8, 10)
  mod <- c(2.2, 3.9, 6.4, 7.5, 10.1)
  all <- goodness_of_fit(obs, mod)
  subset <- goodness_of_fit(obs, mod, metrics = c("kge", "nse", "mse"))
  expect_equal(names(subset), c("kge", "nse", "mse"))
  expect_equal(unname(subset[["kge"]]), unname(all[["kge"]]))
  expect_equal(unname(subset[["nse"]]), unname(all[["nse"]]))
  expect_equal(unname(subset[["mse"]]), unname(all[["mse"]]))
})

test_that("an unknown metric name is rejected and names the offender", {
  expect_error(goodness_of_fit(1:5, 1:5, metrics = "nsee"), "nsee")
})

test_that("mismatched lengths are rejected before reaching C++", {
  expect_error(goodness_of_fit(1:5, 1:4), "same length")
})

test_that("classification_metrics on identical binary labels", {
  obs <- c(1, 0, 1, 1, 0)
  result <- classification_metrics(obs, obs)
  expected_names <- c("accuracy", "precision", "recall", "f1", "specificity",
                       "balanced_accuracy")
  expect_true(all(expected_names %in% names(result)))
  expect_equal(unname(result[["accuracy"]]), 100)
})

test_that("gof_test accepts a corehydro_dist and rejects anything else", {
  x <- c(2.1, 3.4, 1.8, 4.9, 3.3, 2.7, 5.1, 3.9)
  d <- distribution("Normal", c(3.4, 1.1))
  expect_true(gof_test(x, d, "ks") > 0)
  expect_true(gof_test(x, d, "ad") > 0)
  expect_error(gof_test(x, "Normal"), "corehydro_dist")
})

test_that("the information criteria match the closed forms", {
  expect_equal(aic(k = 2, log_likelihood = -100), 204)
  expect_equal(sum(aic_weights(c(246, 248.8, 251.2))), 1)
})

File: corehydropy/tests/test_gof.py
# Binding-specific behaviour for the goodness-of-fit surface. Oracle-value checks live in
# test_fixtures.py, driven by the shared language-neutral fixtures. Mirrors corehydror's
# test-gof.R assertion for assertion.
import pytest

from corehydropy import (
    Distribution,
    aic,
    aic_weights,
    classification_metrics,
    gof_test,
    goodness_of_fit,
)


def test_goodness_of_fit_returns_every_metric_and_subsets_agree_with_the_whole():
    obs = [2, 4, 6, 8, 10]
    mod = [2.2, 3.9, 6.4, 7.5, 10.1]
    every = goodness_of_fit(obs, mod)
    assert len(every) == 17
    assert {"nse", "kge", "rsr", "ve"} <= every.keys()
    assert goodness_of_fit(obs, mod, metrics=["nse"])["nse"] == every["nse"]


def test_goodness_of_fit_accepts_a_bare_metric_string():
    obs = [2, 4, 6, 8, 10]
    mod = [2.2, 3.9, 6.4, 7.5, 10.1]
    every = goodness_of_fit(obs, mod)
    result = goodness_of_fit(obs, mod, metrics="nse")
    assert list(result.keys()) == ["nse"]
    assert result["nse"] == every["nse"]


def test_a_subset_that_excludes_mape_succeeds_on_a_zero_containing_series():
    obs = [0, 2, 4]
    mod = [0.1, 2.1, 3.9]
    with pytest.raises(Exception, match="zero"):
        goodness_of_fit(obs, mod)
    result = goodness_of_fit(obs, mod, metrics="mse")
    assert result["mse"] == pytest.approx(0.01)


def test_a_subset_returns_the_same_values_as_the_full_call_in_the_requested_order():
    obs = [2, 4, 6, 8, 10]
    mod = [2.2, 3.9, 6.4, 7.5, 10.1]
    every = goodness_of_fit(obs, mod)
    subset = goodness_of_fit(obs, mod, metrics=["kge", "nse", "mse"])
    assert list(subset.keys()) == ["kge", "nse", "mse"]
    assert subset["kge"] == every["kge"]
    assert subset["nse"] == every["nse"]
    assert subset["mse"] == every["mse"]


def test_an_unknown_metric_name_is_rejected_and_names_the_offender():
    with pytest.raises(ValueError, match="nsee"):
        goodness_of_fit([1, 2, 3, 4, 5], [1, 2, 3, 4, 5], metrics=["nsee"])


def test_mismatched_lengths_are_rejected_before_reaching_cpp():
    with pytest.raises(ValueError, match="same length"):
        goodness_of_fit([1, 2, 3, 4, 5], [1, 2, 3, 4])


def test_classification_metrics_on_identical_binary_labels():
    obs = [1, 0, 1, 1, 0]
    result = classification_metrics(obs, obs)
    expected_names = {"accuracy", "precision", "recall", "f1", "specificity", "balanced_accuracy"}
    assert expected_names <= result.keys()
    assert result["accuracy"] == 100.0


def test_gof_test_accepts_a_distribution_and_rejects_anything_else():
    x = [2.1, 3.4, 1.8, 4.9, 3.3, 2.7, 5.1, 3.9]
    d = Distribution("Normal", [3.4, 1.1])
    assert gof_test(x, d, "ks") > 0
    assert gof_test(x, d, "ad") > 0
    with pytest.raises(TypeError, match="Distribution"):
        gof_test(x, "Normal")


def test_the_information_criteria_match_the_closed_forms():
    assert aic(k=2, log_likelihood=-100) == 204
    assert sum(aic_weights([246, 248.8, 251.2])) == pytest.approx(1.0)
